combined forcing, structures, states and results improvements

DirkEilander opened the pull request:
Refactor code:
* separated many low-level methods into utils.py and plots.py
* splitted `setup_basemaps` into multiple smaller methods

added support for:
* structures: sfincs.thd & sfincs.weir
* results: sfincs_map.nc & sfincs_his.nc
* states: sfincs.restart
* forcing: sfincs.precip

SfincsModel:
* new methods `setup_p_forcing_from_grid` and `setup_p_forcing` with support for spatial uniform precip, fixes #18
* deprecated `setup_p_gridded`
* new `setup_merge_topobathy` to merge a new topo/bathymetry dataset with the basemap DEM, fixes #22 
* new `setup_mask` and `setup_bounds` methods to setup the sfincs mask file
* new `setup_structures` method
* thd/weir files are read/written as part of staticgeoms, fixes #19
* added `read_states` / `write_states` methods, fixes #17 
* added `read_results` method
* new `update_spatial_attrs` and `get_spatial_attrs` (previously part of `read_staticmaps`)
* save bzs/bzd & dis/src only as GeoDataArray at forcing and do not copy the locations at staticgeoms.

Small improvements
* sort src/bnd files on x_dim for comparibility between OS
* staticmaps are by default saved (and read) in S->N orientation as this matches the SFINCS better.
* sfincs.inp leading when reading forcing, fixes #20 
* improved forcing plots with precip bar plot and limited to simulation time, fixes #21 
* removed pinned dask version
* updated examples, fixes #23 

TODO

- [x] update docs

Leynse reviewed (commented):
Mainly had a look at the new utils.py, 3 questions:
utils.py > L388 in write time-series, what is done with w (time?), made an integer time/why?

utils.py > L434, write_structures, is value 0.6 user definable still?

utils.py > L651, maybe nice to make the value of hmax > 0 user definable? sometimes you want to put that the same as huthresh, or evenhigher like 0.5m

DirkEilander commented:
Thanks @Leynse !

> utils.py > L388 in write time-series, what is done with w (time?), made an integer time/why?

w is the required width of the time column, for formatting only. I've added a comment to explain. 

> utils.py > L434, write_structures, is value 0.6 user definable still?

Yes, through the columns of the geopandas.GeoDataFrame. I've added more explanation to the docstring.

> utils.py > L651, maybe nice to make the value of hmax > 0 user definable? sometimes you want to put that the same as huthresh, or evenhigher like 0.5m

Good suggestion. I've added the option. This can also be afterwards of course.
